File: tester/interpreter/instruction.py
import re
from myerrors import XMLformatError


class Instruction:
    def __init__(self, name):
        self.name = name
        self.arg_types = []
        self.arg_contents = []
        self.arg_count = 0

    def check_inst_without_args(self):
        if self.name not in ["POPFRAME", "RETURN", "PUSHFRAME", "CREATEFRAME", "BREAK"]:
            raise XMLformatError("Error 32: Unknown instruction or valid instruction without needed arguments!"
                                 " Got (" + self.name + ")")

    def check_argument_count(self, count):
        if self.name in ["DEFVAR", "POPS", "PUSHS", "WRITE", "EXIT", "DPRINT", "CALL", "LABEL", "JUMP"] and count == 1 \
            or self.name in ["MOVE", "INT2CHAR", "STRLEN", "TYPE", "NOT", "READ"] and count == 2 \
            or self.name in ["ADD", "SUB", "MUL", "IDIV", "LT", "GT", "EQ", "AND", "OR", "STRI2INT",
                             "CONCAT", "GETCHAR", "SETCHAR", "JUMPIFEQ", "JUMPIFNEQ"] and count == 3:
            return
        else:
            raise XMLformatError("Error 32: Received wrong ammount of arguments in instruction "
                + self.name + "!")

    def arg_append(self, arg_type, arg_content, arg_count):
        self.arg_count = arg_count
        self.__check_corresponding_args(arg_type)
        arg_value = self.__check_arg_type(arg_type, arg_content)
        self.arg_contents.append(arg_value)
        self.arg_types.append(arg_type)

    def __check_corresponding_args(self, arg_type):
        # instruction expects var
        if self.name in ["DEFVAR", "POPS"]:
            if arg_type == "var":
                return
        # instruction expects symb
        elif self.name in ["PUSHS", "WRITE", "EXIT", "DPRINT"]:
            if arg_type in ["string", "bool", "int", "nil", "var"]:
                return
        # instruction expects label
        elif self.name in ["CALL", "LABEL", "JUMP"]:
            if arg_type == "label":
                return
        # instruction expects var,symb
        elif self.name in ["MOVE", "INT2CHAR", "STRLEN", "TYPE", "NOT"]:
            if arg_type == "var" and self.arg_count == 1 or\
               arg_type in ["string", "bool", "int", "nil", "var"] and self.arg_count == 2:
                return
            else:
                print("test")
        # instruction expects var, type
        elif self.name == "READ":
            if arg_type == "var" and self.arg_count == 1 or\
               arg_type == "type" and self.arg_count == 2:
                return
        # instruction expects var,symb, symb
        elif self.name in ["ADD", "SUB", "MUL", "IDIV", "LT", "GT", "EQ", "AND", "OR", "STRI2INT",
                           "CONCAT", "GETCHAR", "SETCHAR"]:
            if arg_type == "var" and self.arg_count == 1 or\
               arg_type in ["string", "bool", "int", "nil", "var"] and self.arg_count == 2 or \
               arg_type in ["string", "bool", "int", "nil", "var"] and self.arg_count == 3:
                return
        # instruction expects label, symb, symb
        elif self.name in ["JUMPIFEQ", "JUMPIFNEQ"]:
            if arg_type == "label" and self.arg_count == 1 or\
               arg_type in ["string", "bool", "int", "nil", "var"] and self.arg_count == 2 or \
               arg_type in ["string", "bool", "int", "nil", "var"] and self.arg_count == 3:
                return
        # unknown instruction
        else:
            raise XMLformatError("Error 32: Received unknown instruction \"" + self.name + "\"!")
        raise XMLformatError("Error 32: Received wrong argument type for arg" + str(self.arg_count) + " in instruction "
                             + self.name + "!")

    def __check_arg_type(self, arg_type, arg_content):
        if arg_type == "type":
            if re.match('^(int|string|bool)$', arg_content):
                return arg_content
        elif arg_type == "var":
            if re.match('^((LF|GF|TF)@([a-z]|[A-Z]|[0-9]|_|-|\$|&|%|\*|!|\?)+)$', arg_content):
                return arg_content
        elif arg_type == "label":
            if re.match('^(([a-z]|[A-Z]|[0-9]|_|-|\$|&|%|\*|!|\?)+)$', arg_content):
                return arg_content
        elif arg_type == "string":
            if re.match('^((\\\\[0-9]{3}|[^#\\\\\s])*)$', arg_content):
                arg_content = self.__transform_escape_sequences(arg_content)
                return arg_content
        elif arg_type == "int":
            if re.match('^([0-9]|-|\+)+$', arg_content):
                return int(arg_content)
        elif arg_type == "bool":
            if re.match('^(true|false)$', arg_content):
                if arg_content == "true":
                    return True
                elif arg_content == "false":
                    return False
        elif arg_type == "nil":
            if re.match('^nil$', arg_content):
                return "nil"
        else:
            raise XMLformatError("Error 32: The arg" + str(self.arg_count) + " in instruction " +
                                 self.name +
                                 " is not valid! Unknown data type \"" + arg_type + "\"")
        raise XMLformatError("Error 32: The content of arg" + str(self.arg_count) + " in instruction " +
                             self.name + " is not valid! Got \"" + arg_content + "\"")

    def __transform_escape_sequences(self, string):
        espace_sequences = re.findall('\\\\\d{3}', string)
        if espace_sequences:
            for seq in espace_sequences:
                seq_list = list(seq)
                number = seq_list[1:]
                number = int("".join(number))
                transformed_seq = chr(number)
                string = string.replace(seq, transformed_seq)
        return string

File: src/parser/instruction.php
<?php

class Instruction{

    public $partitions;
    public $argCount;
    public $insCount;

    public function __construct(){
        $this->insCount = 1;
    }

    public function checkHeader($Instruct, $header){
        $Instruct->checkComment($header);
        $header = trim($header);
        if($header != ".IPPcode20"){
            exit(21);
        }
    }

    public function checkArgCount($expectedCount){
        if($this->argCount == $expectedCount){
            return;
        }
        else{
            fwrite(STDERR, "WRONG AMMOUNT OF ARGUMENTS\n");
            exit(23);
        }
    }

    public function correctVarSyntax($inputLine){
        $inputLine = trim($inputLine);
        preg_match('/^([LF,GF,TF].@.[a-z,A-Z,0-9,\_,\-,\$,\&,\%,\*,\!,\?]*)$/', $inputLine, $outputArray);
        if(empty($outputArray)){
            fwrite(STDERR, "Wrong variable syntax!\n");
            exit(23);
        }
    }

    public function correctSymbSyntax($inputLine, &$symbParts){
        $inputLine = trim($inputLine);
        preg_match('/^(int|bool|string|nil)@.*/', $inputLine, $outputArray);
        if(empty($outputArray)){
            $symbParts = explode("@", $inputLine);
            preg_match('/^([LF,GF,TF].@.[a-z,A-Z,0-9,\_,\-,\$,\&,\%,\*,\!,\?]*)$/', $inputLine, $outputArray2);
            if(empty($outputArray2)){
                fwrite(STDERR, "Wrong symbol syntax!\n");
                exit(23);
            }
        }
        else{
            $symbParts = explode("@", $inputLine);
            switch($symbParts[0]){
                case "string":
                    preg_match('/^string@(\\[0-9]{3}|[^\\#])*$/', $inputLine, $outputArray3);
                    if(empty($outputArray3)){
                        fwrite(STDERR, "Wrong symbol syntax!\n");
                        exit(23);
                    }
                break;
                case "int":
                    preg_match('/^int@([^\\#])*$/', $inputLine, $outputArray3);
                    if($symbParts[1] == NULL){
                        fwrite(STDERR, "Empty int!\n");
                        exit(23);
                    }

                    if(empty($outputArray3)){
                        fwrite(STDERR, "Wrong symbol syntax!\n");
                        exit(23);
                    }
                break;
                case "bool":
                    preg_match('/^bool@(true|false)$/', $inputLine, $outputArray3);
                    if(empty($symbParts[1])){
                        fwrite(STDERR, "Empty bool!\n");
                        exit(23);
                    }
                    if(empty($outputArray3)){
                        fwrite(STDERR, "Wrong symbol syntax!\n");
                        exit(23);
                    }
                break;
                case "nil":
                    preg_match('/^nil@nil$/', $inputLine, $outputArray3);
                    if(empty($symbParts[1])){
                        fwrite(STDERR, "Empty nil!\n");
                        exit(23);
                    }
                    if(empty($outputArray3)){
                        fwrite(STDERR, "Wrong symbol syntax!\n");
                        exit(23);
                    }
                break;
                default:
                    fwrite(STDERR, "Unknown datatype!\n");
                    exit(23);
            }
        }
    }

    public function correctLabelSyntax($inputLine){
        preg_match('/^([a-z,A-Z,0-9,\_,\-,\$,\&,\%,\*,\!,\?]*)$/', $inputLine, $outputArray);
        if(empty($outputArray)){
            fwrite(STDERR, "Wrong label syntax!\n");
            exit(23);
        }
    }

    public function correctType($inputLine){
        preg_match('/^(int|string|bool)$/', $inputLine, $outputArray);
        if(empty($outputArray)){
            fwrite(STDERR, "Wrong type syntax!\n");
            exit(23);
        }
    }

    public function parse(&$line, &$Output){
        $argTypes = ["buzna"];
        $symbParts = array(array());
        $this->checkComment($line);
        if(ctype_space($line)){
            return;
        }
        $line = trim($line);
        $this->partitions = explode(" ", $line);
        $this->argCount = count($this->partitions);
        $this->partitions[0] = strtoupper($this->partitions[0]);
        switch ($this->partitions[0]){
            case "MOVE":
            case "NOT":
                $this->checkArgCount(3);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $argTypes = ['var', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "CREATEFRAME":
                $this->checkArgCount(1);
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, NULL);
            break;
            case "PUSHFRAME":
                $this->checkArgCount(1);
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, NULL);
            break;
            case "POPFRAME":
                $this->checkArgCount(1);
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, NULL);
            break;
            case "DEFVAR":
                $this->checkArgCount(2);
                $this->correctVarSyntax($this->partitions[1]);
                $argTypes = ['var'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, NULL);
            break;
            case "CALL":
                $this->checkArgCount(2);
                $this->correctLabelSyntax($this->partitions[1]);
                $argTypes = ['label'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, NULL);
            break;
            case "RETURN":
                $this->checkArgCount(1);
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "PUSHS":
                $this->checkArgCount(2);
                $this->correctSymbSyntax($this->partitions[1], $symbParts[0]);
                $argTypes = ['symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "POPS":
                $this->checkArgCount(2);
                $this->correctVarSyntax($this->partitions[1]);
                $argTypes = ['var'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "ADD":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "SUB":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "MUL":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "IDIV":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "LT":
            case "GT":
            case "EQ":
            case "AND":
            case "OR":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "INT2CHAR":
                $this->checkArgCount(3);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $argTypes = ['var', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "STRI2INT":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "READ":
                $this->checkArgCount(3);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctType($this->partitions[2]);
                $argTypes = ['var', 'type'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, NULL);
            break;
            case "WRITE":
                $this->checkArgCount(2);
                $this->correctSymbSyntax($this->partitions[1], $symbParts[0]);
                $argTypes = ['symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "CONCAT":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "STRLEN":
                $this->checkArgCount(3);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $argTypes = ['var', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "GETCHAR":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "SETCHAR":
                $this->checkArgCount(4);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['var', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "TYPE":
                $this->checkArgCount(3);
                $this->correctVarSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $argTypes = ['var', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "LABEL":
                $this->checkArgCount(2);
                $this->correctLabelSyntax($this->partitions[1]);
                $argTypes = ['label'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "JUMP":
                $this->checkArgCount(2);
                $this->correctLabelSyntax($this->partitions[1]);
                $argTypes = ['label'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "JUMPIFEQ":
            case "JUMPIFNEQ":
                $this->checkArgCount(4);
                $this->correctLabelSyntax($this->partitions[1]);
                $this->correctSymbSyntax($this->partitions[2], $symbParts[0]);
                $this->correctSymbSyntax($this->partitions[3], $symbParts[1]);
                $argTypes = ['label', 'symb', 'symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "EXIT":
                $this->checkArgCount(2);
                $this->correctSymbSyntax($this->partitions[1], $symbParts[0]);
                $argTypes = ['symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "DPRINT":
                $this->checkArgCount(2);
                $this->correctSymbSyntax($this->partitions[1], $symbParts[0]);
                $argTypes = ['symb'];
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            case "BREAK":
                $this->checkArgCount(1);
                $Output->appendInstruction($this->partitions, $this->insCount, $this->argCount, $argTypes, $symbParts);
            break;
            default:
                if ((trim($this->partitions[0])) == "") {
                    break;
                }
                fwrite(STDERR, "Unknown instruction!".$this->partitions[0]."\n");
                exit(22);
        }
   }

   public function checkComment(&$line){
        if(($charsToNeedle = strpos($line, '#')) !== false){
            $line = substr($line, 0, strpos($line, "#"));
            $line = trim($line);
            $line = $line."\n";
        }
   }
}
?>
